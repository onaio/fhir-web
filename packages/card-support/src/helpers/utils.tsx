/** Function to download data to a file
 *
 * @param {string} data - data to be written to file
 * @param {string} filename - name of the file to be saved
 * @param {string} type - MIME type for the file
 */
export const downloadFile = (data: string, filename: string, type: string) => {
  const file = new Blob([data], { type });
  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
  if (window.navigator.msSaveOrOpenBlob) {
    // IE10+
    window.navigator.msSaveOrOpenBlob(file, filename);
  } else {
    // Others
    const docElement = document.createElement('a');
    const url = URL.createObjectURL(file);
    docElement.href = url;
    docElement.download = filename;
    document.body.appendChild(docElement);
    docElement.click();
    setTimeout(() => {
      document.body.removeChild(docElement);
      window.URL.revokeObjectURL(url);
    }, 0);
  }
};
