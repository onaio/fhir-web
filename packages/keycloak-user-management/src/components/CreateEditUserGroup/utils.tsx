import { KeycloakService } from '@opensrp/keycloak-service';
import { store } from '@opensrp/store';
import { sendErrorNotification, sendSuccessNotification } from '@opensrp/notifications';
import lang, { Lang } from '../../lang';
import { KEYCLOAK_URL_ASSIGNED_ROLES, KEYCLOAK_URL_USER_GROUPS } from '../../constants';
import { KeycloakUserRole } from '../../ducks/userRoles';
import { fetchKeycloakUserGroups, KeycloakUserGroup } from '../../ducks/userGroups';

/**
 * Fetch available, assigned or effective roles
 *
 * @param {string} groupId - user group id
 * @param {string} keycloakBaseURL - keycloak API base URL
 * @param {string} roleMappingEndpoint - keycloak endpoint for fetching assigned, available or effective roles
 * @param {Function} setRolesAction method to set state for selected actions
 * @param {Lang} langObj - the language object
 */
export const fetchRoleMappings = async (
  groupId: string,
  keycloakBaseURL: string,
  roleMappingEndpoint: string,
  setRolesAction: (role: KeycloakUserRole[]) => void,
  langObj: Lang = lang
) => {
  const keycloakService = new KeycloakService(
    `${KEYCLOAK_URL_USER_GROUPS}/${groupId}${roleMappingEndpoint}`,
    keycloakBaseURL
  );

  return await keycloakService
    .list()
    .then((response: KeycloakUserRole[]) => {
      setRolesAction(response);
    })
    .catch((_: Error) => {
      sendErrorNotification(langObj.ERROR_OCCURED);
    });
};

/**
 * Remove assigned roles
 *
 * @param {string} groupId - user group id
 * @param {string} keycloakBaseURL - keycloak API base URL
 * @param {KeycloakUserRole[]} allRoles - an array of all realm roles
 * @param {string[]} rolesToRemove - list of role ids
 * @param {Lang} langObj - the language object
 */
export const removeAssignedRoles = async (
  groupId: string,
  keycloakBaseURL: string,
  allRoles: KeycloakUserRole[],
  rolesToRemove: string[],
  langObj: Lang = lang
) => {
  const data: KeycloakUserRole[] = [];
  rolesToRemove.forEach((roleId: string) => {
    const roleObj = allRoles.find((role: KeycloakUserRole) => role.id === roleId);
    data.push(roleObj as KeycloakUserRole);
  });
  const keycloakService = new KeycloakService(
    `${KEYCLOAK_URL_USER_GROUPS}/${groupId}${KEYCLOAK_URL_ASSIGNED_ROLES}`,
    keycloakBaseURL
  );

  return await keycloakService
    .update(data, null, 'DELETE')
    .then(() => {
      sendSuccessNotification(langObj.ROLES_UPDATED_SUCCESSFULLY);
    })
    .catch((_: Error) => {
      sendErrorNotification(langObj.ERROR_OCCURED);
    });
};

/**
 * Set assigned roles
 *
 * @param {string} groupId - user group id
 * @param {string} keycloakBaseURL - keycloak API base URL
 * @param {KeycloakUserRole[]} allRoles - an array of all realm roles
 * @param {string[]} rolesToAdd - list of role ids
 * @param {Lang} langObj - the language object
 */
export const assignRoles = async (
  groupId: string,
  keycloakBaseURL: string,
  allRoles: KeycloakUserRole[],
  rolesToAdd: string[],
  langObj: Lang = lang
) => {
  const data: KeycloakUserRole[] = [];
  rolesToAdd.forEach((roleId: string) => {
    const roleObj = allRoles.find((role: KeycloakUserRole) => role.id === roleId);
    data.push(roleObj as KeycloakUserRole);
  });
  const keycloakService = new KeycloakService(
    `${KEYCLOAK_URL_USER_GROUPS}/${groupId}${KEYCLOAK_URL_ASSIGNED_ROLES}`,
    keycloakBaseURL
  );

  return await keycloakService
    .create(data)
    .then(() => {
      sendSuccessNotification(langObj.ROLES_UPDATED_SUCCESSFULLY);
    })
    .catch((_: Error) => {
      sendErrorNotification(langObj.ERROR_OCCURED);
    });
};

/**
 * Fetch single user group
 *
 * @param {string} groupId -
 * @param {string} keycloakBaseURL - keycloak API base URL
 * @param {Function} dispatch method to dispatch action to store
 * @param {Lang} langObj - the language object
 */
export const fetchSingleGroup = async (
  groupId: string,
  keycloakBaseURL: string,
  dispatch: typeof store.dispatch,
  langObj: Lang = lang
) => {
  const keycloakService = new KeycloakService(KEYCLOAK_URL_USER_GROUPS, keycloakBaseURL);

  await keycloakService
    .read(groupId)
    .then((response: KeycloakUserGroup) => {
      dispatch(fetchKeycloakUserGroups([response]));
    })
    .catch((_: Error) => {
      sendErrorNotification(langObj.ERROR_OCCURED);
    });
};
