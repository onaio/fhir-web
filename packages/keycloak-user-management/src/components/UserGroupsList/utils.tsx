import { KeycloakService } from '@opensrp/keycloak-service';
import { sendErrorNotification } from '@opensrp/notifications';
import { KEYCLOAK_URL_USER_GROUPS } from '../../constants';
import { KeycloakUserGroup } from '../../ducks/userGroups';
import { UserGroupMembers } from '.';
import lang, { Lang } from '../../lang';

// data loader utils for user group detail view

/** Function to fetch group members from keycloak
 *
 * @param {string} groupId - user group id
 * @param {string} baseURL - keycloak base url
 * @param {Function} callback - callback function to set group members from api to state
 * @param {Lang} langObj - the translation strings lookup
 */
export const loadGroupMembers = async (
  groupId: string,
  baseURL: string,
  callback: (members: UserGroupMembers[]) => void,
  langObj: Lang = lang
) => {
  const serve = new KeycloakService(`${KEYCLOAK_URL_USER_GROUPS}/${groupId}/members`, baseURL);
  return await serve
    .list()
    .then((response: UserGroupMembers[]) => {
      callback(response);
    })
    .catch(() => sendErrorNotification(langObj.ERROR_OCCURED));
};

/**
 * Function to fetch single group details from keycloak
 *
 * @param {string} groupId - user group id
 * @param {string} baseURL - keycloak base url
 * @param {Function} callback - callback function to set group members from api to state
 * @param {Lang} langObj - the translation strings lookup
 */
export const loadGroupDetails = async (
  groupId: string,
  baseURL: string,
  callback: (userGroups: KeycloakUserGroup) => void,
  langObj: Lang = lang
) => {
  const serve = new KeycloakService(KEYCLOAK_URL_USER_GROUPS, baseURL);
  return await serve
    .read(groupId)
    .then((response: KeycloakUserGroup) => {
      callback(response);
    })
    .catch(() => sendErrorNotification(langObj.ERROR_OCCURED));
};
