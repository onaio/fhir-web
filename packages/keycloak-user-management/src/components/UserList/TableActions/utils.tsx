import { notification } from 'antd';
import { KeycloakService } from '@opensrp/keycloak-service';
import { removeKeycloakUsers } from '../../../ducks/user';
import { KEYCLOAK_URL_USERS, USER_DELETED_SUCCESSFULLY, ERROR_OCCURED } from '../../../constants';

/**
 * Delete keycloak user
 *
 * @param {Function} removeKeycloakUsersCreator - remove users action creator
 * @param {string} accessToken - access token
 * @param {string} keycloakBaseURL - keycloak api base URL
 * @param {string} userId - id of user to be deleted
 * @param {Function} isLoadingCallback - callback function that sets loading state
 * @returns {void}
 */
export const deleteUser = (
  removeKeycloakUsersCreator: typeof removeKeycloakUsers,
  accessToken: string,
  keycloakBaseURL: string,
  userId: string,
  isLoadingCallback: (loading: boolean) => void
): void => {
  const serviceDelete = new KeycloakService(
    accessToken,
    `${KEYCLOAK_URL_USERS}/${userId}`,
    keycloakBaseURL
  );
  serviceDelete
    .delete()
    .then(() => {
      removeKeycloakUsersCreator();
      isLoadingCallback(true);
      notification.success({
        message: `${USER_DELETED_SUCCESSFULLY}`,
        description: '',
      });
    })
    .catch((_: Error) => {
      notification.error({
        message: `${ERROR_OCCURED}`,
        description: '',
      });
    });
};
