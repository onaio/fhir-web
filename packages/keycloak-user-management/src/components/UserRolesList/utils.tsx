import { sendErrorNotification } from '@opensrp/notifications';
import { KeycloakService } from '@opensrp/keycloak-service';
import { store } from '@opensrp/store';
import { KEYCLOAK_URL_USER_ROLES } from '../../constants';
import { ERROR_OCCURED } from '../../lang';
import { fetchKeycloakUserRoles, KeycloakUserRole } from '../../ducks/userRoles';

/**
 * Fetch all realm roles
 *
 * @param {string} keycloakBaseURL - keycloak API base URL
 * @param {Function} dispatch method to dispatch action to store
 */
export const fetchAllRoles = async (keycloakBaseURL: string, dispatch: typeof store.dispatch) => {
  const keycloakService = new KeycloakService(KEYCLOAK_URL_USER_ROLES, keycloakBaseURL);

  await keycloakService
    .list()
    .then((response: KeycloakUserRole[]) => {
      dispatch(fetchKeycloakUserRoles(response));
    })
    .catch((_: Error) => {
      sendErrorNotification(ERROR_OCCURED);
    });
};
