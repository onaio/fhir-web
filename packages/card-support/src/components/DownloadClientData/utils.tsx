import Papaparse from 'papaparse';
import { OpenSRPService } from '@opensrp/server-service';
import { DownloadClientDataFormFields } from '../DownloadClientData';
import { Dispatch, SetStateAction } from 'react';
import { OPENSRP_URL_CLIENT_SEARCH, APPLICATION_CSV } from '../../constants';
import { Client } from '../../ducks/clients';
import { downloadFile } from '../../helpers/utils';
import { ParsedHierarchyNode } from '@opensrp/location-management';
import { sendErrorNotification } from '@opensrp/notifications';
import { Dictionary } from '@onaio/utils';
import lang, { Lang } from '../../lang';
/* eslint-disable @typescript-eslint/camelcase */

/** interface for user assignment response */
export interface UserAssignment {
  organizationIds: number[];
  jurisdictions: string[];
  plans: string[];
}

/** interface for CSV entries */
export interface ClientCSVEntry {
  id: string;
  dob: string;
  first_name: string;
  last_name: string;
  gender: string;
  facility_of_registration: string;
  date_of_registration: string;
  card_status: string;
  card_status_last_update: string;
}

/**
 * Handle form date change
 *
 * @param {string} range - selected date range
 * @param {Dispatch<SetStateAction<[string, string]>>} setCardOrderDate - state dispatcher
 */
export const handleCardOrderDateChange = (
  range: [string, string],
  setCardOrderDate: Dispatch<SetStateAction<[string, string]>>
): void => {
  setCardOrderDate(range);
};

/**
 * Handle download client data form submission
 *
 * @param values - sumitted form values
 * @param accessToken - OPENSRP API access token
 * @param opensrpBaseURL - OPENSRP API base URL
 * @param serviceClass - OPENSRP service class
 * @param locations location hierarchy
 * @param setSubmitting - method to set form `isSubmitting` status
 * @param langObj - tthe lang object
 */
export const submitForm = (
  values: DownloadClientDataFormFields,
  accessToken: string,
  opensrpBaseURL: string,
  serviceClass: typeof OpenSRPService,
  locations: ParsedHierarchyNode[],
  setSubmitting: (isSubmitting: boolean) => void,
  langObj: Lang = lang
): void => {
  const { clientLocation, cardStatus, cardOrderDate } = values;

  if (!clientLocation) {
    return;
  }

  setSubmitting(true);
  const startDate = cardOrderDate[0];
  const endDate = cardOrderDate[1];
  const endPoint = `${OPENSRP_URL_CLIENT_SEARCH}`;
  const serve = new serviceClass(accessToken, opensrpBaseURL, endPoint);
  let params: Dictionary = {
    locationIds: clientLocation,
  };

  if (cardStatus) {
    params = {
      ...params,
      attribute: `card_status:${cardStatus}`,
    };
  }

  serve
    .list(params)
    .then((clients: Client[]) => {
      const location = getLocationDetails(locations, clientLocation);
      const entries: ClientCSVEntry[] = clients
        .filter((client: Client) => !!client.identifiers.zeir_id)
        .filter((client: Client) => {
          const dateCreated = new Date(client.dateCreated.split('T')[0]);

          return (
            dateCreated.getTime() >= new Date(startDate).getTime() &&
            dateCreated.getTime() <= new Date(endDate).getTime()
          );
        })
        .map((client: Client) => {
          return {
            id: client.identifiers.zeir_id,
            dob: formatDDMMYYY(new Date(client.birthdate.split('T')[0])),
            first_name: client.firstName,
            last_name: client.lastName,
            gender: client.gender,
            facility_of_registration: location ? location.title : '',
            date_of_registration: formatDDMMYYY(new Date(client.dateCreated.split('T')[0])),
            card_status: client.attributes.card_status,
            card_status_last_update: formatDDMMYYY(
              new Date(client.attributes.card_status_date.split('T')[0])
            ),
          };
        });

      setSubmitting(false);

      if (entries.length) {
        createCsv(entries, buildCSVFileName(location ? location.title : '', startDate, endDate));
      } else {
        sendErrorNotification(langObj.NO_DATA_FOUND);
      }
    })
    .catch((_: Error) => {
      setSubmitting(false);
      sendErrorNotification(langObj.ERROR_OCCURRED);
    });
};

/**
 * Format a Date object as dd MM YYYY
 *
 * @param {Date} date - date object
 * @param {string} delimiter - delimiter to use to separate the different parts
 * @returns {string} formatted date
 */
export const formatDDMMYYY = (date: Date, delimiter = '/'): string => {
  //eslint-disable-next-line @typescript-eslint/restrict-plus-operands
  return `${('0' + date.getDate()).slice(-2)}${delimiter}${('0' + (date.getMonth() + 1)).slice(
    -2
  )}${delimiter}${date.getFullYear()}`;
};

/**
 * Build the CSV file name
 *
 * @param {string} location - name of the location filter
 * @param {string} startDate - start date of the range filter
 * @param {string} endDate - end date of the range filter
 * @returns {string} CSV file name
 */
export const buildCSVFileName = (location: string, startDate: string, endDate: string): string => {
  return `Children_list_${location}_${formatDDMMYYY(new Date(), '_')}_(${formatDDMMYYY(
    new Date(startDate),
    '-'
  )} - ${formatDDMMYYY(new Date(endDate), '-')})`;
};

// Create csv data for the selected jurisdiction hieracrhy
export const createCsv = (entries: ClientCSVEntry[], fileName: string): void => {
  const csv: string = Papaparse.unparse(entries, {
    header: true,
  });
  // Export csv file
  downloadFile(csv, `${fileName}.csv`, APPLICATION_CSV);
};

/**
 * Get location from the location hierarchy
 *
 * @param {ParsedHierarchyNode} locations location hierarchy
 * @param {string} locationId id of location to return
 * @returns {ParsedHierarchyNode | null} location if found, null if not found
 */
export const getLocationDetails = (
  locations: ParsedHierarchyNode[],
  locationId: string
): ParsedHierarchyNode | null => {
  let found = false;
  let i = 0;
  let location = null;

  while (!found && i < locations.length) {
    const node = locations[i];

    if (node.id && node.id === locationId) {
      found = true;
      location = node;
    } else if (node.children) {
      location = getLocationDetails(node.children, locationId);

      if (location) {
        found = true;
      }
    }

    i += 1;
  }

  return location;
};
