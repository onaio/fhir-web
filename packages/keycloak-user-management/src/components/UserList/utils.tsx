import { Dictionary } from '@onaio/utils';
import lang, { Lang } from '../../lang';
import { Column } from '@opensrp/react-utils';
import { TableData } from '.';

/**
 * Get table columns for user list
 *
 * @param {Dictionary} sortedInfo - applied sort
 * @param {Lang} langObj - translations object lookup
 * @returns {Column<TableData>[]} - an array of table columns
 */
export const getTableColumns = (
  sortedInfo?: Dictionary,
  langObj: Lang = lang
): Column<TableData>[] => {
  const headerItems: string[] = [
    // langObj.USERNAME,
    langObj.EMAIL,
    langObj.FIRST_NAME,
    langObj.LAST_NAME,
  ];
  const dataElements: Column<TableData>[] = [];
  const fields: string[] = ['username', 'email', 'firstName', 'lastName'];

  fields.forEach((field: string, index: number) => {
    dataElements.push({
      title: headerItems[index],
      dataIndex: field as keyof TableData,
      key: field as keyof TableData,
      sorter: (a: Dictionary, b: Dictionary) => {
        if (a[field] > b[field]) {
          return -1;
        } else if (a[field] < b[field]) {
          return 1;
        }
        return 0;
      },
      sortOrder: sortedInfo && sortedInfo.columnKey === field && sortedInfo.order,
      ellipsis: true,
    });
  });

  return dataElements;
};
