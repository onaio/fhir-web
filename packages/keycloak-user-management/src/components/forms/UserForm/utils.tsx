import { notification } from 'antd';
import { history } from '@onaio/connected-reducer-registry';
import { KeycloakUser } from '@opensrp/store';
import { Dictionary } from '@onaio/utils';
import { Dispatch, SetStateAction } from 'react';
import { KeycloakService } from '@opensrp/keycloak-service';
import {
  KEYCLOAK_URL_USERS,
  URL_ADMIN,
  KEYCLOAK_URL_REQUIRED_USER_ACTIONS,
  ERROR_OCCURED,
} from '../../../constants';

/**
 * Handle form submission
 *
 * @param {Dictionary} values - form values
 * @param {string} accessToken - keycloak API access token
 * @param {string} keycloakBaseURL - keycloak API base URL
 * @param {KeycloakService} keycloakServiceClass - keycloak API service class
 * @param {Function} setSubmitting - method to set submission status
 * @param {string} userId - keycloak user id, required when editing a user
 */
export const submitForm = (
  values: Partial<KeycloakUser>,
  accessToken: string,
  keycloakBaseURL: string,
  keycloakServiceClass: typeof KeycloakService,
  setSubmitting: (isSubmitting: boolean) => void,
  userId?: string
): void => {
  setSubmitting(true);

  if (userId) {
    const serve = new keycloakServiceClass(
      accessToken,
      `${KEYCLOAK_URL_USERS}/${userId}`,
      keycloakBaseURL
    );
    serve
      .update(values)
      .then(() => {
        setSubmitting(false);
        history.push(URL_ADMIN);
        notification.success({
          message: 'User edited successfully',
          description: '',
        });
      })
      .catch((_: Error) => {
        setSubmitting(false);
        notification.error({
          message: ERROR_OCCURED,
          description: '',
        });
      });
  } else {
    const serve = new keycloakServiceClass(accessToken, KEYCLOAK_URL_USERS, keycloakBaseURL);
    serve
      .create(values)
      .then(() => {
        setSubmitting(false);
        history.push(URL_ADMIN);
        notification.success({
          message: 'User created successfully',
          description: '',
        });
      })
      .catch((_: Error) => {
        setSubmitting(false);
        notification.error({
          message: ERROR_OCCURED,
          description: '',
        });
      });
  }
};

/** interface user action */
export interface UserAction {
  alias: string;
  name: string;
  providerId: string;
  enabled: boolean;
  defaultAction: boolean;
  priority: number;
  config: Dictionary;
}

/**
 * Fetch keycloak user action options
 *
 * @param {string} accessToken - keycloak API access token
 * @param {string} keycloakBaseURL - keycloak API base URL
 * @param {Function} setUserActionOptions - method to set state for selected actions
 * @param {KeycloakService} keycloakServiceClass - keycloak API service class
 */
export const fetchRequiredActions = (
  accessToken: string,
  keycloakBaseURL: string,
  setUserActionOptions: Dispatch<SetStateAction<UserAction[]>>,
  keycloakServiceClass: typeof KeycloakService
): void => {
  const keycloakService = new keycloakServiceClass(
    accessToken,
    KEYCLOAK_URL_REQUIRED_USER_ACTIONS,
    keycloakBaseURL
  );

  keycloakService
    .list()
    .then((response: UserAction[]) => {
      setUserActionOptions(
        response.filter((action: UserAction) => action.alias !== 'terms_and_conditions')
      );
    })
    .catch((_: Error) => {
      notification.error({
        message: ERROR_OCCURED,
        description: '',
      });
    });
};
