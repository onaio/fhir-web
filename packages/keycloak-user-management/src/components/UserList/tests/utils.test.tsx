import * as fixtures from '../../forms/UserForm/tests/fixtures';
import { getTableColumns } from '../utils';
/* eslint-disable @typescript-eslint/camelcase */

describe('components/UserList/utils/getTableColumns', () => {
  it('builds table columns correctly', () => {
    const removeUsersMock = jest.fn();
    const isLoadingCallbackMock = jest.fn();
    const accessToken = 'sometoken';
    const keycloakBaseURL =
      'https://keycloak-stage.smartregister.org/auth/admin/realms/opensrp-web-stage';

    expect(
      getTableColumns(removeUsersMock, accessToken, keycloakBaseURL, isLoadingCallbackMock, {
        user_id: fixtures.keycloakUser.id,
      })
    ).toMatchSnapshot('table columns');
  });
});
