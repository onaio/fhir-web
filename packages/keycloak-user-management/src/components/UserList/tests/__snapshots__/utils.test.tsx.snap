// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/UserList/utils/getDataFilters builds table columns correctly: table columns 1`] = `
Array [
  Object {
    "dataIndex": "username",
    "ellipsis": true,
    "filteredValue": null,
    "filters": Array [
      Object {
        "text": "mwalimu",
        "value": "mwalimu",
      },
      Object {
        "text": "ona",
        "value": "ona",
      },
    ],
    "key": "username",
    "onFilter": [Function],
    "sortOrder": undefined,
    "sorter": [Function],
    "title": "Username",
  },
  Object {
    "dataIndex": "email",
    "ellipsis": true,
    "filteredValue": null,
    "filters": Array [
      Object {
        "text": "dubdabasoduba@gmail.com",
        "value": "dubdabasoduba@gmail.com",
      },
      Object {
        "text": undefined,
        "value": undefined,
      },
    ],
    "key": "email",
    "onFilter": [Function],
    "sortOrder": undefined,
    "sorter": [Function],
    "title": "Email",
  },
  Object {
    "dataIndex": "firstName",
    "ellipsis": true,
    "filteredValue": null,
    "filters": Array [
      Object {
        "text": "Benjamin",
        "value": "Benjamin",
      },
      Object {
        "text": "Ona",
        "value": "Ona",
      },
    ],
    "key": "firstName",
    "onFilter": [Function],
    "sortOrder": undefined,
    "sorter": [Function],
    "title": "First Name",
  },
  Object {
    "dataIndex": "lastName",
    "ellipsis": true,
    "filteredValue": null,
    "filters": Array [
      Object {
        "text": "Mwalimu",
        "value": "Mwalimu",
      },
      Object {
        "text": "kenya",
        "value": "kenya",
      },
    ],
    "key": "lastName",
    "onFilter": [Function],
    "sortOrder": undefined,
    "sorter": [Function],
    "title": "Last Name",
  },
  Object {
    "dataIndex": "actions",
    "key": "Actions",
    "render": [Function],
    "title": "Actions",
  },
]
`;
